File: src/main/java/com/intgrah/algorithms/heap/StrictFibonacciHeap.java
package com.intgrah.algorithms.heap;

import com.intgrah.algorithms.list.Deletable;
import com.intgrah.algorithms.list.DoublyLinkedList;

import java.util.ArrayList;
import java.util.Comparator;

public class StrictFibonacciHeap<K> extends DecreasableHeap<K> {

    private final ArrayList<Rank> rankList = new ArrayList<>();
    private final DoublyLinkedList<Node> queue = new DoublyLinkedList<>();
    private final DoublyLinkedList<Rank> activeRootsReady = new DoublyLinkedList<>();
    private final DoublyLinkedList<Node> oneNodeLossReady = new DoublyLinkedList<>();
    private final DoublyLinkedList<Rank> twoNodeLossReady = new DoublyLinkedList<>();
    private Node root;

    { rankList.add(new Rank()); }

    public StrictFibonacciHeap(Comparator<K> ord) { super(ord); }

    @Override
    public Decreasable pushRef(K k) {
        Node x = new Node();
        Box b = new Box(k);
        b.bind(x);
        if (isEmpty()) {
            root = x;
        } else {
            if (ord.compare(k, root.box.key) <= 0) {
                Node r = root;
                link(r, root = x);
                queue.pushFront(r);
//                root = x;
            } else {
                link(x, root);
                queue.pushFront(x);
            }
            activeRootReduction();
            rootDegreeReduction();
        }
        return b;
    }

    private void link(Node n, Node p) {
        if (n.parentRef != null)
            n.parentRef.delete();
        n.parent = p;
        if (n.isActive)
            n.parentRef = p.activeChildren.pushFront(n);
        else if (!n.isLinkable())
            n.parentRef = p.passiveChildren.pushFront(n);
        else
            n.parentRef = p.linkableChildren.pushFront(n);
    }

    private boolean activeRootReduction() {
        if (activeRootsReady.isEmpty())
            return false;
        Rank r = activeRootsReady.getFront();
        Node x = r.popActiveRoot();
        Node y = r.popActiveRoot();
        x.rankRef = null;
        y.rankRef = null;
        if (ord.compare(x.box.key, y.box.key) > 0) {
            Node z = x;
            x = y;
            y = z;
        }
        Node w = y.parent;
        link(y, x);
        if (w.isLinkable())
            link(w, w.parent);
        x.setRank(x.rank + 1);
        if (!x.linkableChildren.isEmpty()) {
            Node z = x.linkableChildren.getBack();
            link(z, root);
        } else if (!x.passiveChildren.isEmpty()) {
            Node z = x.passiveChildren.getBack();
            link(z, root);
        }
        return true;
    }

    private boolean rootDegreeReduction() {
        if (root.linkableChildren.size() < 3)
            return false;
        ArrayList<Node> xyz = new ArrayList<>();
        xyz.add(root.linkableChildren.popBack());
        xyz.add(root.linkableChildren.popBack());
        xyz.add(root.linkableChildren.popBack());
        xyz.sort(Comparator.comparing(n -> n.box.key, ord));
        Node x = xyz.get(0);
        Node y = xyz.get(1);
        Node z = xyz.get(2);
        x.parentRef = null;
        y.parentRef = null;
        z.parentRef = null;
        x.isActive = y.isActive = true;
        link(z, y);
        link(y, x);
        link(x, root);
        x.setRank(1);
        return true;
    }

    @Override
    public K getMin() {
        assert !isEmpty();
        return root.box.key;
    }

    @Override
    public void deleteMin() {
        assert !isEmpty();
        Node r = root;
        root = null;
        Node x = null;
        for (Node c : r.activeChildren)
            if (x == null || ord.compare(c.box.key, x.box.key) < 0)
                x = c;
        for (Node c : r.passiveChildren)
            if (x == null || ord.compare(c.box.key, x.box.key) < 0)
                x = c;
        for (Node c : r.linkableChildren)
            if (x == null || ord.compare(c.box.key, x.box.key) < 0)
                x = c;
        if (x == null)
            return;
        x.isActive = false;
        if (x.rankRef != null)
            rankList.get(x.rank).removeNode(x);
        x.parentRef.delete();
        x.parentRef = null;
        x.parent = null;
        for (Node c : r.activeChildren)
            link(c, x);
        for (Node c : r.passiveChildren)
            link(c, x);
        for (Node c : r.linkableChildren)
            link(c, x);
        for (Node c : x.activeChildren) {
            rankList.get(c.rank).removeNode(c);
            rankList.get(c.rank).addActiveRoot(c);
            c.setLoss(0);
        }
        root = x;
        for (int i = 0; i < Math.min(2, queue.size()); i++) {
            Node y = queue.getFront();
            queue.pushBack(y);
            if (y.passiveChildren.size() + y.linkableChildren.size() >= 2) {
                for (int j = 0; j < 2; j++) {
                    Node z;
                    if (!y.linkableChildren.isEmpty())
                        z = y.linkableChildren.popFront();
                    else
                        z = y.passiveChildren.popFront();
                    z.parentRef = null;
                    link(z, root);
                }
                if (y.isActive && !y.isActiveRoot())
                    y.setLoss(y.loss + 2);
            }
        }
        if (!twoNodeLossReduction())
            oneNodeLossReduction();
        while (activeRootReduction() || rootDegreeReduction())
            assert true;
    }

    private boolean twoNodeLossReduction() {
        if (twoNodeLossReady.isEmpty())
            return false;
        Rank r = twoNodeLossReady.getFront();
        Node x = r.popTwoNodeLossNode();
        Node y = r.popTwoNodeLossNode();
        x.rankRef = null;
        y.rankRef = null;
        if (ord.compare(x.box.key, y.box.key) > 0) {
            Node z = x;
            x = y;
            y = z;
        }
        Node z = y.parent;
        link(y, x);
        y.setLoss(0);
        if (x != z) {
            x.setRank(x.rank + 1);
            x.setLoss(0);
            z.setRank(z.rank - 1);
            if (!z.isActiveRoot())
                z.setLoss(z.loss + 1);
        }
        return true;
    }

    private boolean oneNodeLossReduction() {
        if (oneNodeLossReady.isEmpty())
            return false;
        Node x = oneNodeLossReady.getFront();
        Node y = x.parent;
        x.setLoss(0);
        link(x, root);
        Rank record = rankList.get(x.rank);
        record.removeNode(x);
        record.addActiveRoot(x);
        y.setRank(y.rank - 1);
        if (!y.isActiveRoot())
            y.setLoss(y.loss + 1);
        return true;
    }

    @Override
    public boolean isEmpty() { return root == null; }

    @Override
    public void clear() { root = null; }

    private class Rank {

        final DoublyLinkedList<Node> activeRoots = new DoublyLinkedList<>();
        final DoublyLinkedList<Node> twoNodeLossPairs = new DoublyLinkedList<>();
        Deletable activeRootsReadyNode;
        Deletable twoNodeLossReadyNode;

        void removeNode(Node x) {
            if (x.rankRef == null)
                return;
            x.rankRef.delete();
            x.rankRef = null;
            refreshActiveRootsReady();
            refreshTwoNodeLossReady();
        }

        void refreshActiveRootsReady() {
            if (activeRoots.size() < 2) {
                if (activeRootsReadyNode != null) {
                    activeRootsReadyNode.delete();
                    activeRootsReadyNode = null;
                }
            } else {
                if (activeRootsReadyNode == null)
                    activeRootsReadyNode = activeRootsReady.pushFront(this);
            }
        }

        void refreshTwoNodeLossReady() {
            if (twoNodeLossPairs.size() < 2) {
                if (twoNodeLossReadyNode != null) {
                    twoNodeLossReadyNode.delete();
                    twoNodeLossReadyNode = null;
                }
            } else {

                if (twoNodeLossReadyNode == null)
                    twoNodeLossReadyNode = twoNodeLossReady.pushFront(this);
            }
        }

        void addActiveRoot(Node x) {
            x.rankRef = activeRoots.pushFront(x);
            refreshActiveRootsReady();
        }

        Node popActiveRoot() {
            Node x = activeRoots.popFront();
            x.rankRef = null;
            refreshActiveRootsReady();
            return x;
        }

        void addTwoNodeLossNode(Node x) {
            x.rankRef = twoNodeLossPairs.pushFront(x);
            refreshTwoNodeLossReady();
        }

        Node popTwoNodeLossNode() {
            Node x = twoNodeLossPairs.popFront();
            x.rankRef = null;
            refreshTwoNodeLossReady();
            return x;
        }

    }


    private class Box extends Decreasable {

        Node ref;

        Box(K k) { super(k); }

        @Override
        public void decreaseKey(K k) {
            assert ord.compare(k, key) <= 0;
            key = k;
            Node x = ref;
            if (x == root)
                return;
            if (ord.compare(key, root.box.key) < 0) {
                root.box.bind(ref);
                bind(root);
            }
            Node y = x.parent;
            x.setLoss(0);
            link(x, root);
            if (x.isActiveRoot()) {
                rankList.get(x.rank).removeNode(x);
                rankList.get(x.rank).addActiveRoot(x);
            }
            if (y.isActive) {
                if (x.isActive)
                    y.setRank(y.rank - 1);
                if (!y.isActiveRoot())
                    y.setLoss(y.loss + 1);
            } else if (y.isLinkable()) {
                link(y, y.parent);
            }
            if (!twoNodeLossReduction())
                oneNodeLossReduction();
            for (int i = 0; i < 6; i++) {
                activeRootReduction();
                rootDegreeReduction();
            }
        }

        void bind(Node x) {
            x.box = this;
            ref = x;
        }

    }


    private class Node {

        final DoublyLinkedList<Node> activeChildren = new DoublyLinkedList<>();
        final DoublyLinkedList<Node> passiveChildren = new DoublyLinkedList<>();
        final DoublyLinkedList<Node> linkableChildren = new DoublyLinkedList<>();
        Box box;
        Node parent = null;
        Deletable parentRef;
        boolean isActive = false;
        int rank = 0;
        Deletable rankRef;
        int loss = 0;
        Deletable lossRef;

        void setLoss(int l) {
            rankList.get(rank).removeNode(this);
            loss = l;
            if (loss == 0) {
                if (lossRef != null) {
                    lossRef.delete();
                    lossRef = null;
                }
            }
            if (loss == 1)
                rankList.get(rank).addTwoNodeLossNode(this);
            else if (loss >= 2) {
                if (lossRef != null)
                    lossRef = oneNodeLossReady.pushFront(this);
            }
        }

        void setRank(int r) {
            rankList.get(rank).removeNode(this);
            rank = r;
            if (rank == rankList.size())
                rankList.add(new Rank());
            if (isActiveRoot())
                rankList.get(rank).addActiveRoot(this);
            else if (loss == 1)
                rankList.get(rank).addTwoNodeLossNode(this);
        }

        boolean isActiveRoot() { return isActive && !parent.isActive; }

        boolean isLinkable() {
            return !isActive && activeChildren.isEmpty() && passiveChildren.isEmpty();
        }

    }

}

File: heap/strict_fibonacci_heap.py
from collections import deque
from __future__ import annotations

class Box[K]:
    def __init__(self, k: K) -> None:
        self.key = k
        self.node: Node = None
        
    def bind(self, n: Node) -> None:
        n.box = self
        self.node = n


class Node[K]:
    def __init__(self) -> None:
        self.active = list[Node]()
        self.passive = list[Node]()
        self.linkable = list[Node]()

        self.box: Box = None
        self.parent: Node[K] | None = None
        self.isActive = False
        self.rank = 0
        self.loss = 0
        
    def detach(self) -> None:
        assert self.parent is not None
        if self.isActive:
            self.parent.active.remove(self)
        elif not self.isLinkable():
            self.parent.passive.remove(self)
        else:
            self.parent.linkable.remove(self)
        self.parent = None
            
    def attach(self, p: Node[K]) -> None:
        assert self.parent is None
        self.parent = p
        if self.isActive:
            p.active.append(self)
        elif not self.isLinkable():
            p.passive.append(self)
        else:
            p.linkable.append(self)
            
    def reattach(self) -> None:
        p = self.parent
        self.detach()
        self.attach(p)

    def isActiveRoot(self):
        return self.isActive and not self.parent.isActive

    def isLinkable(self):
        return not self.isActive and self.activeChildren.isEmpty() and self.passiveChildren.isEmpty()

    

class Rank[K]:
    def __init__(self) -> None:
        self.activeRoots: list[Node[K]] = []
        self.twoNodeLossPairs: list[Node[K]] = []


class StrictFibonacciHeap[K]:
    def __init__(self) -> None:
        self.rank_list: list[Rank] = [Rank()]
        self.queue = deque[Node] = deque()
        # self.activeRootsReady = set[Rank]()
        # self.oneNodeLossReady = set[Node]()
        # self.twoNodeLossReady = set[Rank]()
        # self.nodes = set[Node]()
        self.root: Node = None

    def push(self, k: K):
        x = Node[K]()
        b = Box(k)
        b.bind(x)
        # self.nodes.add(x) ############################
        if self.is_empty():
            self.root = x
        else:
            if k <= self.root.box.key:
                r = self.root
                self.root = x
                r.attach(x)
                self.queue.append(r)
            else:
                x.attach(self.root)
                self.queue.pushFront(x)
            self.activeRootReduction()
            self.rootDegreeReduction()

        return b


    def activeRootReduction(self) -> bool:
        # Check if available
        ready = [r for r in self.rank_list if len(r.activeRoots) >= 2]
        if not ready:
            return False
        r = ready[0]
        x = r.activeRoots.pop()
        y = r.activeRoots.pop()
        if x.box.key > y.box.key:
            x, y = y, x
        w = y.parent
        y.detach()
        y.attach(x)
        if w.isLinkable():
            w.reattach()
        x.setRank(x.rank + 1)
        if len(x.linkable) > 0:
            z = x.linkable[-1]
            z.detach()
            z.attach(self.root)
        elif not x.passiveChildren.isEmpty():
            z = x.passiveChildren[-1]
            z.detach()
            z.attach(self.root)
        return True

    def rootDegreeReduction(self) -> bool:
        if len(self.root.linkable) < 3:
            return False
        xyz = [self.root.linkable.pop() for _ in range(3)]
        xyz.sort(key=lambda n: n.box.key)
        x, y, z = xyz
        x.detach()
        y.detach()
        z.detach()
        x.isActive = True
        y.isActive = True
        z.attach(y)
        y.attach(x)
        x.setRank(1)
        x.attach(self.root)
        return True

    def oneNodeLossReduction(self) -> bool:
        if not self.oneNodeLossReady.isEmpty():
            return False
        x = self.oneNodeLossReady.getFront()
        Node y = x.parent
        x.setLoss(0)
        link(x, root)
        Rank record = rankList.get(x.rank)
        record.removeNode(x)
        record.addActiveRoot(x)
        y.setRank(y.rank - 1)
        if (!y.isActiveRoot())
            y.setLoss(y.loss + 1)
        return true

    def get_min(self) -> K:
        assert not self.is_empty()
        return self.root.box.key

    def delete_min(self) -> None:
        assert not self.is_empty()
        r = self.root
        self.root = None
        x: Node = None
        for c in r.active:
            if x is None or c.box.key < x.box.key:
                x = c
        for c in r.passive:
            if x is None or c.box.key < x.box.key:
                x = c
        for c in r.linkable:
            if x is None or c.box.key < x.box.key:
                x = c
        if x == None:
            return
        x.isActive = False
        if x.rankRef is not None:
            rankList.get(x.rank).removeNode(x)
        x.parentRef.delete()
        x.parentRef = None
        x.parent = None
        for c in r.active:
            c.attach(x)
        for c in r.passive:
            c.attach(x)
        for c in r.linkable:
            c.attach(x)
        for c in x.activeChildren:
            rankList.get(c.rank).removeNode(c)
            rankList.get(c.rank).addActiveRoot(c)
            c.setLoss(0)
        self.root = x
        for i in range(min(2, len(self.queue))):
            y = self.queue.pop()
            self.queue.append(y)
            if (y.passiveChildren.size() + y.linkableChildren.size() >= 2):
                for j in range(2):
                    if not y.linkableChildren.isEmpty():
                        z = y.linkableChildren.popFront()
                    else:
                        z = y.passiveChildren.popFront()
                    z.parentRef = None
                    link(z, root)
                if y.isActive and not y.isActiveRoot():
                    y.setLoss(y.loss + 2)
        if not twoNodeLossReduction():
            oneNodeLossReduction()
        while activeRootReduction() or rootDegreeReduction():
            assert True

    def twoNodeLossReduction(self) -> bool:
        if twoNodeLossReady.isEmpty():
            return False
        Rank r = twoNodeLossReady.getFront()
        Node x = r.popTwoNodeLossNode()
        Node y = r.popTwoNodeLossNode()
        x.rankRef = None
        y.rankRef = None
        if (ord.compare(x.box.key, y.box.key) > 0) {
            Node z = x
            x = y
            y = z
        }
        Node z = y.parent
        link(y, x)
        y.setLoss(0)
        if (x != z) {
            x.setRank(x.rank + 1)
            x.setLoss(0)
            z.setRank(z.rank - 1)
            if (!z.isActiveRoot())
                z.setLoss(z.loss + 1)
        }
        return true
    
    def decreaseKey(self, b: Box, k: K) -> None:
        assert k <= box.key
        b.key = k
        x = self.node
            if x == root:
                return
            if key < root.box.key:
                root.box.bind(ref)
                bind(root)

            Node y = x.parent
            x.setLoss(0)
            link(x, root)
            if (x.isActiveRoot()) {
                rankList.get(x.rank).removeNode(x)
                rankList.get(x.rank).addActiveRoot(x)
            }
            if (y.isActive) {
                if (x.isActive)
                    y.setRank(y.rank - 1)
                if (!y.isActiveRoot())
                    y.setLoss(y.loss + 1)
            } else if (y.isLinkable()) {
                link(y, y.parent)
            }
            if (!twoNodeLossReduction())
                oneNodeLossReduction()
            for i in range(6):
                activeRootReduction()
                rootDegreeReduction()


    def is_empty(self) -> bool:
        return self.root == None

    def clear(self) -> None:
        self.root = None
